### PR TITLE
Migrate to Witty

### PR DESCRIPTION
**Pull request by jvff**
## Motivation

Witty allows specifying WIT interfaces using Rust code, and supports different host Wasm runtimes, allowing us to update the `wasmer`, `wasmtime` and `wit-bindgen` dependencies to newer versions.

## Proposal

Migrate to use Witty in the host side (and in the `linera-wasm-test-runner`) and use the generated WIT files with a newer version of `wit-bindgen` in the `linera-sdk`.

## Test Plan

This is a major refactor, and existing tests should catch any regressions.

## Release Plan

This changes the WIT interface, so:
- Need to bump the major/minor version number in the next release of the crates.

## Links

- [reviewer checklist](https://github.com/linera-io/linera-protocol/blob/main/CONTRIBUTING.md#reviewer-checklist)

### CODE REVIEW

File: linera-execution/src/lib.rs [hunk 1]
```diff
@@ -42,22 +36,30 @@ use linera_base::{
...
+    },
+    execution::ExecutionStateView,
+    policy::ResourceControlPolicy,
+    resources::{ResourceController, ResourceTracker},
+    system::{
+        SystemExecutionError, SystemExecutionStateView, SystemMessage, SystemOperation,
+        SystemQuery, SystemResponse,
+    },
+};
 pub use crate::runtime::{ContractSyncRuntime, ServiceSyncRuntime};
```
Comment (ma2bd) [inline linera-execution/src/lib.rs:63]: What's the rationale for `self` vs `crate`?

Comment (jvff) [reply linera-execution/src/lib.rs:63]: Good question, I think it's just a style choice I had on auto-pilot :sweat_smile: I'llchange it to `crate`.

**Review by ma2bd — APPROVED** [review]
Yay!

File: linera-sdk/Cargo.toml [hunk 2]
```diff
@@ -51,11 +58,10 @@ linera-chain = { workspace = true, features = ["metrics"] }
 linera-core = { workspace = true, features = ["test", "metrics", "wasmer"] }
 linera-execution = { workspace = true, features = ["fs", "metrics", "wasmer"] }
 linera-storage = { workspace = true, features = ["metrics", "wasmer"] }
-linera-witty = { workspace = true, features = ["test"] }
+linera-witty = { workspace = true, features = ["wasmer", "test"] }
```
Comment (afck) [inline linera-sdk/Cargo.toml:61]: Isn't `wasmer` optional?

File: linera-execution/src/lib.rs [hunk 3]
```diff
@@ -42,23 +36,31 @@ use linera_base::{
...
 use serde::{Deserialize, Serialize};
-pub use system::{
-    SystemExecutionError, SystemExecutionStateView, SystemMessage, SystemOperation, SystemQuery,
-    SystemResponse,
-};
 use thiserror::Error;
+
+#[cfg(with_testing)]
+pub use crate::applications::ApplicationRegistry;
 #[cfg(all(with_testing, any(with_wasmer, with_wasmtime)))]
```
Comment (afck) [inline linera-execution/src/lib.rs:44]:
```suggestion
#[cfg(all(with_testing, with_wasm_runtime))]
```

File: linera-execution/src/wasm/wasmer.rs [hunk 4]
```diff
@@ -85,33 +55,49 @@ where
...
+            .user_data_mut()
+            .runtime_mut()
             .consume_fuel(self.initial_fuel - remaining_fuel)
     }
 }
 
 /// Type representing a running [Wasmer](https://wasmer.io/) service.
-pub struct WasmerServiceInstance {
+pub struct WasmerServiceInstance<Runtime> {
     /// The application type.
```
Comment (afck) [inline linera-execution/src/wasm/wasmer.rs:98]:
```suggestion
```

File: linera-sdk/src/views/aliases.rs [hunk 5]
```diff
@@ -0,0 +1,61 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Aliases for views using the [`ViewStorageContext`].
+
+use super::ViewStorageContext;
+
+/// An alias to [`linera_views::collection_view::ByteCollectionView`] that uses the
+/// WebAssembly specific [`ViewStorageContext`].
```
Comment (afck) [inline linera-sdk/src/views/aliases.rs:9]: (Nit: I'd write "WebAssembly-specific" with a hyphen everywhere, but I'm not sure.)

File: linera-sdk/src/views/system_api_stubs.rs [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+use linera_views::{
+    batch::Batch,
+    common::{ContextFromStore, KeyValueStore, ReadableKeyValueStore, WritableKeyValueStore},
+    views::ViewError,
+};
+
+const MESSAGE: &str = "Attempt to call a contract system API when not running as a Wasm guest";
+
+/// A type to interface with the key value storage provided to applications.
```
Comment (afck) [inline linera-sdk/src/views/system_api_stubs.rs:16]:
```suggestion
/// A type to interface with the key-value storage provided to applications.
```

**Review by afck — APPROVED** [review]
Beautiful! :star_struck:

File: linera-sdk/Cargo.toml (hunk 2 above)
Comment (jvff) [reply linera-sdk/Cargo.toml:61]: Yes, I meant to use `wasmtime` here, which is needed by the `linera-wasm-test-runner`. Maybe I should just add it as a `required-features` for the binary instead? :thinking:

**Pull request merged**